torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -0,0 +1,406 @@
+#include "ProcessGroupGloo.hpp"
+
+#include <gloo/allreduce_halving_doubling.h>
+#include <gloo/broadcast_one_to_all.h>
+#include <gloo/rendezvous/context.h>
+#include <gloo/transport/tcp/device.h>
+
+#define GENERATE_ALL_TYPES(type, func, args...)        \
+  switch (type) {                                      \
+    case ::at::ScalarType::Float:                      \
+      func<float>(args);                               \
+      break;                                           \
+    case ::at::ScalarType::Double:                     \
+      func<double>(args);                              \
+      break;                                           \
+    case ::at::ScalarType::Half:                       \
+      func<gloo::float16>(args);                       \
+      break;                                           \
+    case ::at::ScalarType::Char:                       \
+      func<int8_t>(args);                              \
+      break;                                           \
+    case ::at::ScalarType::Byte:                       \
+      func<uint8_t>(args);                             \
+      break;                                           \
+    case ::at::ScalarType::Int:                        \
+      func<int32_t>(args);                             \
+      break;                                           \
+    case ::at::ScalarType::Long:                       \
+      func<int64_t>(args);                             \
+      break;                                           \
+    default:                                           \
+      throw std::runtime_error("Invalid scalar type"); \
+  }
+
+namespace c10d {
+
+using KeyType = AlgorithmKey;
+using EntryType = std::unique_ptr<AlgorithmEntry>;
+
+namespace {
+
+// Wrap c10d store as Gloo store
+class GlooStore : public ::gloo::rendezvous::Store {
+ public:
+  GlooStore(const std::shared_ptr<::c10d::Store>& store) : store_(store) {}
+
+  void set(const std::string& key, const std::vector<char>& value) override {
+    std::vector<unsigned char> tmp(value.begin(), value.end());
```

Comment:
`unsigned char` doesn't need to be the same type as `uint8_t` in the compiler, and it will cause a compile error if this happens.

Reply:
I'll change it back. I did a copy to an uint8_t vector before which was quite verbose, and using unsigned char as a proxy made made this a bunch easier.

Reply:
Hmm turns out I could swap out unsigned char for uint8_t after all. I could have sworn this caused a bunch of compilation errors before.